### DISCUSSION TITLE
Unable to use positional arguments in the decorator for database_sync_to_async

### DISCUSSION THREAD

**Discussion by Kevv-J** [django/channels#1849] [Help!] [answered]
Hi,
I've been using the sync_to_async wrapper from asgiref but wanted a method that would cleanly close old connections and was redirected to use the database_sync_to_async. 

Earlier 

`@sync_to_async(thread_sensitive=settings.THREAD_SENSITIVE)`
`def foo():`

Now 

`@database_sync_to_async(thread_sensitive=settings.THREAD_SENSITIVE)`
`def foo():`
It gives me the following error:

File "[file-path]", line 27, in [class-name]
    @database_sync_to_async(thread_sensitive=settings.THREAD_SENSITIVE)
TypeError: __init__() missing 1 required positional argument: 'func'

using the following versions:
asgiref==3.4.1
channels==3.0.4

Are positional arguments supported in the decorator?

**Comment by carltongibson** [comment, marked as answer]
Perhaps not with the decorator syntax... — but I'd have to dig in to look. 

I'd imagine it would work as:

```
def foo():
    ...

foo = @database_sync_to_async(foo, thread_sensitive=settings.THREAD_SENSITIVE)
```

**BUT** `thread_sensitive` is specifically for database connections... AFAICS it should never be `False`.

**Reply by Kevv-J** [reply]
Thanks for the quick reply.

The approach you mentioned is what I ended up trying out although it would be convenient to have it work through the decorator.(maybe I can put in a feature request if it is feasible).

Why do you mention that thread_sensitive should never be false? Unless I'm missing something big, isn't having multiple threads run the database queries going to be more efficient in a high load system.

**Reply by carltongibson** [reply]
That's not (exactly) what thread_sensitive means… Rather, it requires that database queries are made on the same thread as the DB connection was created on. It'll break otherwise.

**Reply by Kevv-J** [reply]
We are using Django with channels and running the application with ASGI (Daphne).  As the application is using the ASYNC approach, essentially it runs as a single-threaded application and with thread_sensitive=true, pretty much only one DB connection is ever created. with the application having to server large number of requests, reusing just one DB connection might be a bottleneck and that’s why we were experimenting with thread_sensitive=false. If thread_sensitive=true is the way to go, can you suggest how to have a bit more parallelism when running this application?

**Reply by carltongibson** [reply]
Upgrade to Django 4.0, which introduced per-request contexts for thread sensitive actions.

**Reply by Kevv-J** [reply]
Hey, @carltongibson. I'd like to explore this feature in Django 4.0 and looked into documentation for the same but couldn't find anything related to per-request contexts for thread sensitive actions in 4.0 that are different from 3.2. Could you point me in the right direction?

**Reply by carltongibson** [reply]
@Kevv-J It's an implementation detail... But see https://github.com/django/django/commit/36fa071d6ebd18a61c4d7f1b5c9d17106134bd44
